### PR TITLE
Show confirmation dialog when there are unsaved changes

### LINKED ISSUE #1571: Missing confirm to exit unsaved binding changes
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Steps to reproduce 🕹

Steps:

1. Open https://master--toolpad.mui.com/_toolpad/app/cld06azhx0001k59tpypy3aoy/pages/mu03mbo
2. Select the Text element
3. Click on "Update value binding"
4. Make a change
5. Press <kbd>Escape</kbd>

Same behavior for:

- custom components editor
- page module editor

### Current behavior 😯

No confirmation to exit.

### Expected behavior 🤔

Confirmation before exiting.

### Context 🔦

I lost some of my changes because of this behavior. I press escape to hide this popup:

Screenshot 2023-01-17 at 16 53 16

expecting that it would hide it? But instead, it exited the binding modal. This is how I found this regression (I assume it's one). And yes, it would be great for an escape to not trigger the exit action in this case, as it does with this different popup:

Screenshot 2023-01-17 at 16 54 35

### Your environment 🌎

v0.0.35

### PR DESCRIPTION
**Pull request by apedroferreira**
Closes https://github.com/mui/mui-toolpad/issues/1571

The code component part of this PR can be seen/reviewed isolated here: https://github.com/mui/mui-toolpad/pull/1628
We can review and merge that part first if it makes things easier.

- [x] Confirm when leaving code component editor with unsaved changes (improvements)
- [x] Confirm when leaving query editor with unsaved changes, unless "Cancel" is pressed
- [x] Confirm when leaving page module dialog unsaved changes, unless "Cancel" is pressed
- [x] Confirm when leaving page parameters dialog with unsaved changes, unless "Close" is pressed
- [x] Confirm when leaving bindings dialog with unsaved changes, unless "Cancel" is pressed

All except the code component flow use the same `useEditorUnsavedChangesConfirm` hook.

Preview: https://toolpad-pr-1618.onrender.com/

### CODE REVIEW

File: packages/toolpad-app/src/toolpad/hooks/useUnsavedChangesConfirm.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+export default function useUnsavedChangesConfirm({
+  hasUnsavedChanges,
+  onClose,
+}: UseUnsavedChangesConfirmInput): UseUnsavedChangesConfirmPayload {
+  const handleCloseWithUnsavedChanges = React.useCallback(() => {
+    if (hasUnsavedChanges) {
+      // eslint-disable-next-line no-alert
+      const hasConfirmed = window.confirm(
+        'You have unsaved changes. Are you sure you want to navigate away?\nAll changes will be discarded.',
```
Comment (Janpot) [inline packages/toolpad-app/src/toolpad/hooks/useUnsavedChangesConfirm.ts:22]:
```suggestion
        'You have unsaved changes. Are you sure you want to navigate away? All changes will be discarded when you press "Ok".',
```

Maybe the newline looks a bit odd?

Screenshot 2023-02-20 at 17 23 19

**Review by Janpot — APPROVED** [review]
Works really well. I'm having some doubts about the cancel button behavior, but no idea what would be the best UX.

Maybe the newline in the dialog looks a bit off?

**Comment by apedroferreira** [comment]
> Works really well. I'm having some doubts about the cancel button behavior, but no idea what would be the best UX.
> 
> Maybe the newline in the dialog looks a bit off?

We can always easily change it in the future.

**Comment by oliviertassinari** [comment]
> I'm having some doubts about the cancel button behavior, but no idea what would be the best UX.

Playing a bit with the UX, I personally faced no frustration on this, so far. I will report back here if I find something.

**Comment by oliviertassinari** [comment]
@apedroferreira I might have found a bug:

1. Open https://master--toolpad.mui.com/_toolpad/app/cl9my2gkr0001q89lt82h56ey/pages/q9198sj
2. 

https://user-images.githubusercontent.com/3165635/220422687-f27c1f90-28e6-440a-9ed6-0167b6df2c04.mov

It asks me confirm when I exit even if I did no changes.

**Comment by apedroferreira** [comment]
> @apedroferreira I might have found a bug on this PR. It asks me to confirm my exit when I use Escape even if I did no changes.
> 
> 1. Open https://master--toolpad.mui.com/_toolpad/app/cl9my2gkr0001q89lt82h56ey/pages/q9198sj
> 2. Follow:

Maybe because of auto-formatting there's a difference between the initial value and the value on save... Should be an easy fix, thanks for reporting this, will add a fix in a bit.

**Pull request merged**